rustls/src/msgs/handshake.rs

```diff
@@ -228,100 +228,130 @@ impl TlsListElement for SignatureScheme {
-            }
-            Self::IpAddress(r) => r.encode(bytes),
-            Self::Unknown(r) => r.encode(bytes),
+        let server_name_list = LengthPrefixedBuffer::new(Self::SIZE_LEN, bytes);
+
+        if let ServerNamePayload::SingleDnsName(dns_name) = self {
+            ServerNameType::HostName.encode(server_name_list.buf);
+            let name_slice = dns_name.as_ref().as_bytes();
+            (name_slice.len() as u16).encode(server_name_list.buf);
+            server_name_list
+                .buf
+                .extend_from_slice(name_slice);
         }
     }
-}
 
-#[derive(Clone, Debug)]
-pub struct ServerName {
-    pub(crate) typ: ServerNameType,
-    pub(crate) payload: ServerNamePayload,
-}
+    fn read(r: &mut Reader<'a>) -> Result<Self, InvalidMessage> {
+        let mut found = None;
 
-impl Codec<'_> for ServerName {
-    fn encode(&self, bytes: &mut Vec<u8>) {
-        self.typ.encode(bytes);
-        self.payload.encode(bytes);
-    }
+        let len = Self::SIZE_LEN.read(r)?;
+        let mut sub = r.sub(len)?;
 
-    fn read(r: &mut Reader<'_>) -> Result<Self, InvalidMessage> {
-        let typ = ServerNameType::read(r)?;
+        while sub.any_left() {
+            let typ = ServerNameType::read(&mut sub)?;
+
+            let payload = match typ {
+                ServerNameType::HostName => HostNamePayload::read(&mut sub)?,
+                _ => {
+                    // Consume remainder of extension bytes.  Since the length of the item
+                    // is an unknown encoding, we cannot continue.
+                    sub.rest();
+                    break;
+                }
+            };
 
-        let payload = match typ {
-            ServerNameType::HostName => ServerNamePayload::read_hostname(r)?,
-            _ => ServerNamePayload::Unknown(Payload::read(r).into_owned()),
-        };
+            match payload {
+                // "The ServerNameList MUST NOT contain more than one name of
+                // the same name_type." - RFC6066
+                HostNamePayload::HostName(_) | HostNamePayload::IpAddress(_) => {
+                    if found.is_some() {
+                        warn!("Illegal SNI extension: duplicate host_name received");
+                        return Err(InvalidMessage::InvalidServerName);
+                    }
+                }
+            }
 
-        Ok(Self { typ, payload })
+            match payload {
+                HostNamePayload::HostName(dns_name) => {
+                    found = Some(Self::SingleDnsName(dns_name.to_owned()));
+                }
+
+                // Skip over hostname items that are actually IP addresses.
+                HostNamePayload::IpAddress(_invalid) => {
+                    warn!(
+                        "Illegal SNI extension: ignoring IP address presented as hostname ({:?})",
+                        _invalid
+                    );
+                    found = Some(Self::IpAddress);
+                }
+            }
```

Comment:
I feel like merging these two `match`es might be more readable? Also potentially hoist the `found.is_some()` into the match arm pattern rather than nesting it?